### DISCUSSION TITLE
Correct behavior for client no ACK

### DISCUSSION THREAD

**Discussion by orangepineapple** [ANLAB-KAIST/KENSv3#157] [Project Q&A 2. Transport Layer (TCP)]
In the 3 way handshake the client is supposed to return an ACK to the server. At what point is the client added to the backlog of the listening socket? 

1) Only after the ACK is returned, then during that time what if the backlog is filled, do we just drop the client? Even though we sent the client and ACK
2) Added to backlog after the SYN request, then do we just drop the client if the server called accept before the ACK is returned?

**Comment by Pusnow** [comment]
2 is enough for KENS.



Actually it varies across implementations.

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/Documentation/man/listen.2.txt#L60-L64

**Reply by orangepineapple** [reply]
So what if the server calls accept on this socket before the ack from the client side is received? Do we just assume the client will always ACK back? Or do we just drop the connection?

**Reply by Pusnow** [reply]
> So what if the server calls accept on this socket before the ack from the client side is received?

Until the server receives ACK, the connection is in `SYN-RCVD` state, so `accept()` blocks execution. Please review the TCP state diagram.

> Do we just assume the client will always ACK back? Or do we just drop the connection?

If server does not receive ACK packet from a client (retransmission timeout occurs), there are three possibilities: (1) client's ACK packet is dropped, (2) server's SYN/ACK is dropped, or (3) client intentionally does not send ACK packet. The server cannot distinguish the three cases, but to address (2), server should retransmit SYN/ACK packet if retransmission timeout occurs.

(3) is similar to well-known SYN flood DDOS attack. Even though it is out of scope, if you are interested in this, see https://www.cloudflare.com/learning/ddos/syn-flood-ddos-attack/ .

In KENS, clients alway try to send ACK packets.

**Reply by zee-doubleu-ee** [reply]
So, suppose we called listen on a socket but not yet accept. And, we receive sync requests. So, we need to take note of these clients in the backlog but send nothing back to the client. Then, when we call accept on the socket. Take one of the client at a time and try sending syn-ack back to establish the connection. If the client responds with ack the connection is established on our side and accept should return. If not, we send an reset to that client and move on to another one in the backlog?

**Reply by Pusnow** [reply]
You should send SYN/ACK right afther SYN is received. Connection can be established regardless of `accept()`.

If there are established connection before `accept()`, the `accept` can return without blocking. If not, `accept()` should be blocked.

**Reply by orangepineapple** [reply]
In this case, what happens if the clients starts sending data before we accept? From client perspective the connection is established and it can start writing to the socket. But on server side the receiving socket had not been created using accept() yet.

**Reply by Pusnow** [reply]
Since the connection is established, the client can send data to the server, and data is stored in kernel-side (networking-side) buffer.

When `accept()` and server's `read()` are called, the data will be copied to the application's buffer.

**Reply by zee-doubleu-ee** [reply]
Sorry, for replying to this thread but it's also about backlog so,.. In backlog1 test case, the server starts accepting at one second after listen. But, the client in the backlog sends back fin-ack just after the connection is established. In that case, if the server send back a fin-ack to close the connection, the connection would be already closed when the accept is called on server. Then, it won't pass the test case. So, what is the correct behavior?

**Reply by Pusnow** [reply]
The connection is *not* closed until the server application explicitly call `close()`.
 
> In that case, if the server send back a fin-ack to close the connection ...

No, the server only sends `ACK` to the client, and later, if application calls `close()`, then the server sends `FIN` to the client. (See "passive close" in the state transition diagram.)
